### PR TITLE
update neli to 0.6.3

### PR DESCRIPTION
**Pull request by phi-gamma**
API changes:

    - impl_var!() removed in 0.6.0.
    - Trait Nl removed in 0.6.0.
    - Stream{Read,Write}Buffer have been obsoleted.
    - There's an enum IflaInfo for IFLA_INFO_* defines.
    - NlPayload type for Genlmsghdr payload.
    - Track NlSocketHandle API changes (::connect() over ::open()/::bind()).
    - Simplified header deserialization via FromBytesWithInput trait.
    - The asize() member function disappeared.

### CODE REVIEW

**Comment by phi-gamma** [comment]
Hi, I need ``wireguard-uapi`` to work against a more up-to-date ``neli`` so I
had a stab at updating the dependency. Please tell me if you’re interested,
I’ve been using it for about a week now without running into issues.

Regarding formatting, the repo lacks a ``rustfmt.toml`` and with the default
settings rustfmt will touch a number of lines that are already in the repo.
How do you suggest I proceed with this?

**Comment by ernestask** [comment]
Though this adds unnecessary workload on @phi-gamma, so I could just review this as-is.

**Comment by phi-gamma** [comment]
Sorry, I hadn’t seen that earlier PR so no idea how much they differ.
I could look into rebasing on top of it sure but it may take some time
until I get around to it as I’ve got a four weeks vacation coming up.

**Comment by gluxon** [comment]
> Though this adds unnecessary workload on @phi-gamma, so I could just review this as-is.

I'd be okay with both options. Happy to go with your preference @ernestask.

If we decide to merge https://github.com/gluxon/wireguard-uapi-rs/pull/21 first, I can help rebase the changes here on top of the `0.5.3` to reduce the unnecessary workload on @phi-gamma.  Would that be appreciated @phi-gamma?

**Comment by phi-gamma** [comment]
Rebased.

I tried to preserve as much of @ernestask’s work as I could. Those ``TryFrom``
impls in particular are nifty. :)

**Comment by phi-gamma** [comment]
Just bumped the ``neli`` dep to 0.6.2. There seems to be some issue with the CI
which runs the ``enodev`` test that isn’t even on this branch.

**Review by xtexChooser — APPROVED** [review]

**Comment by gluxon** [comment]
> I now rebased the branch on main and adapted the error message. After the library update we’re getting that ENODEV as expected, however from the comment it looks like besides that you’re also interested in a differently worded error message. Could you maybe comment on that?

Your update looks right. We can followup after this PR merges with a better wording. 🙂

To unblock the base PR, I ended up cherry-picking your fix over to https://github.com/gluxon/wireguard-uapi-rs/pull/21/commits/29c04ba8f3108e21b3b66b6cf0fa9e933d7debd0. Thanks!

**Comment by gluxon** [comment]
A few conflicts appeared after https://github.com/gluxon/wireguard-uapi-rs/pull/21 merged. @phi-gamma Would you appreciate my help resolving those conflicts, or would you prefer to do that yourself?

**Comment by phi-gamma** [comment]
Branch adapted, please have a look now!

File: src/linux/socket/wg_socket.rs
Comment (gluxon) [inline src/linux/socket/wg_socket.rs:126]: Do you recall why the turbofish operator here was removed? Was there an compile error, runtime error, or is this just general code cleanup? If there was a compile or runtime error, I'd be curious to investigate a bit more to make sure the change here is expected.

File: src/linux/socket/wg_socket.rs
Comment (gluxon) [inline src/linux/socket/wg_socket.rs:106]: Cool that the `NlTypeWrapper` is no longer needed. Thanks.

File: src/linux/attr.rs
Comment (gluxon) [inline src/linux/attr.rs:33]: The `neli_enum` macro here is cool.

**Comment by phi-gamma** [comment]
> > @@ -132,7 +123,7 @@ impl WgSocket {
>      pub fn set_device(&mut self, device: set::Device) -> Result<(), SetDeviceError> {
>          for nl_message in create_set_device_messages(device, self.family_id)? {
>              self.sock.send(nl_message)?;
> -            self.sock.recv::<Nlmsg, Genlmsghdr<CtrlCmd, CtrlAttr>>()?;
> +            self.sock.recv()?;
> 
> Do you recall why the turbofish operator here was removed?

Not anymore, no. Trying to restore the old line, I see ``rustc``
complain about a missing error conversion trait so maybe it was a
the “lazy fix” to that? It’s been a while so I can’t say for
certain what the motivation was.

**Comment by gluxon** [comment]
@phi-gamma Thanks again for these changes! I was admittedly a bit nervous about this upgrade, but your changes look good. Appreciate the patience.

**Review by gluxon — APPROVED** [review]

**Comment by gluxon** [comment]
Released v3.0.0-rc1. @phi-gamma @ernestask Give a heads up if you see any issues. Otherwise we can release v3.0.0 in a few weeks.

**Comment by phi-gamma** [comment]
> Merged #26 into main.

Great news, thanks a lot!

**Comment by thomasjfox** [comment]
Some feedback on v3.0.0-rc1:

I've taken the previous version 2.0.5, applied the neli 0.5.3 + 0.6.1 patches and diffed the output compared to v3.0.0-rc1. The only noteworthy change was:

Commit https://github.com/gluxon/wireguard-uapi-rs/commit/f595e5d314d5bc1d6f28ce7696c215089d14cfc9
"Iterate over Ifinfomsg attributes once when listing names"

Next I've taken v3.0.0-rc1 and ran integration tests for a custom Wireguard configuration daemon. All tests passed and they do stuff like change allowed IPs or keepalive intervals at runtime.

The new code has been in production for over a week and hums away.

Thanks a lot @gluxon for merging the neli 0.6.x upgrade.

**Comment by gluxon** [comment]
3.0.0 has been released with the changes here! https://github.com/gluxon/wireguard-uapi-rs/releases/tag/v3.0.0

@thomasjfox Thank you so much for those testing notes. It made releasing the large neli updates over the last few months less scary.

**Pull request merged**
